### PR TITLE
improved scaling error reporting and robustness

### PR DESCRIPTION
**Pull request by tjlane**
Addressing #162 -- scaling can break many ways, current code design is fragile.

There was an issue previously in the code design. NaN values used to mark missing data were sliced out, but this happened after the scaling optimizer ran... if the optimizer went off the rails, it would produce new NaNs, creating chaos when they were sliced out.

This PR refactored some of the code so that slicing missing data happens cleanly before this step, which should be more robust.

I also added safeguards on the optimizer to try and stabilize it.

EDIT: 
### Copilot Summary

This pull request significantly improves the robustness, reliability, and test coverage of the scaling logic in `meteor/scale.py`. The main changes include stricter input validation, better handling of edge cases (such as missing or non-finite data), improved numerical stability, and expanded tests to cover new and existing behaviors. Several new constants and error types are introduced to clarify error handling and make the code easier to maintain.

**Key improvements and changes:**

**1. Robustness and Error Handling**
- Introduced new constants (`MAX_SCALE_FACTOR`, `NON_FINITE_RESIDUAL_PENALTY`, `MIN_FRACTION_COMMON_INDICES`) and a dedicated `ScalingError` exception to handle scaling-specific errors more clearly.
- Added input validation to ensure all scale parameters are finite and raise `ScalingError` if not, preventing silent failures and improving debuggability.
- Improved error messages and checks for cases where there are no valid reflections to scale, or when the initial scaling constant is invalid (non-finite or negative), raising `ScalingError` with descriptive messages.

**2. Numerical Stability**
- Clipped the computed scale factors to a maximum absolute value (`MAX_SCALE_FACTOR`) to prevent numerical overflow and ensure all returned values are finite.
- Used `np.nan_to_num` in residual computation to replace any non-finite values with a large penalty, ensuring optimizers receive stable, finite input.

**3. API and Type Improvements**
- Allowed `compute_scale_factors` to accept either a `pd.Index` or a precomputed `(n, 3)` ndarray for Miller indices, improving performance by avoiding redundant conversions in tight loops.
- Improved type annotations for the `least_squares_loss` parameter to clarify expected callable signatures.

**4. Test Suite Expansion and Modernization**
- Added comprehensive new tests for edge cases, including handling of non-finite parameters, overflow in scale factors, missing data (NaNs), and correct error raising for invalid input.
- Updated tests to use `pytest.mark.parametrize` for systematic coverage of different scale modes and input types.
- Removed unnecessary mocking and legacy code from tests, streamlining the test suite.

**5. Code Quality and Maintainability**
- Refactored internal logic for clarity and efficiency, including extracting a helper function `_cast_to_miller_array` and simplifying array handling throughout.
- Ensured all returned arrays are of consistent type and shape, and removed redundant or unreachable code.

These changes make the scaling functionality more robust, easier to debug, and better tested, reducing the likelihood of silent failures in production and improving maintainability.

- Robustness and Error Handling: [[1]](diffhunk://#diff-61a26a7427dd6659136e0d6aebd9e206b0967c489b82ff610f27be469b221936R22-R32) [[2]](diffhunk://#diff-61a26a7427dd6659136e0d6aebd9e206b0967c489b82ff610f27be469b221936R52-R93) [[3]](diffhunk://#diff-61a26a7427dd6659136e0d6aebd9e206b0967c489b82ff610f27be469b221936L168-R253)
- Numerical Stability: [[1]](diffhunk://#diff-61a26a7427dd6659136e0d6aebd9e206b0967c489b82ff610f27be469b221936L89-R121) [[2]](diffhunk://#diff-61a26a7427dd6659136e0d6aebd9e206b0967c489b82ff610f27be469b221936L168-R253)
- API and Type Improvements: [[1]](diffhunk://#diff-61a26a7427dd6659136e0d6aebd9e206b0967c489b82ff610f27be469b221936R52-R93) [[2]](diffhunk://#diff-61a26a7427dd6659136e0d6aebd9e206b0967c489b82ff610f27be469b221936L105-R130)
- Test Suite Expansion and Modernization: [[1]](diffhunk://#diff-aabbf11104ca7a37bd1a12c51b62fdbaec50a068697cafa69c31742a1fc7d5efL356-R472) [[2]](diffhunk://#diff-aabbf11104ca7a37bd1a12c51b62fdbaec50a068697cafa69c31742a1fc7d5efL2)
- Code Quality and Maintainability: [[1]](diffhunk://#diff-61a26a7427dd6659136e0d6aebd9e206b0967c489b82ff610f27be469b221936R52-R93) [[2]](diffhunk://#diff-61a26a7427dd6659136e0d6aebd9e206b0967c489b82ff610f27be469b221936L221-L233)

### CODE REVIEW

**Comment by cvazz** [comment]
Hey, from the current structure, I have a question that ought to be documented in this thread.

You define a ScalingError, but it only triggers if one of the two input datasets has a negative or nan mean. Doesn't this technically qualify as an Error for unclean data -> negative or nan structure factors detected? 

Is now the expected behavior that previous errors will not occur if this check is passed?

I saw `test_scale_maps_recovers_scale_with_partial_nan_inputs` -> would it also pass if there is only a single shared index? An industrious and eager person might suggest raising a warning if a significant (10%, 50% or 90 %?) number of indices are discarded. Then again, not sure how relevant that is.

**Comment by tjlane** [comment]
> Hey, from the current structure, I have a question that ought to be documented in this thread.
> 
> You define a ScalingError, but it only triggers if one of the two input datasets has a negative or nan mean. Doesn't this technically qualify as an Error for unclean data -> negative or nan structure factors detected?
> 
> Is now the expected behavior that previous errors will not occur if this check is passed?
> 
> I saw `test_scale_maps_recovers_scale_with_partial_nan_inputs` -> would it also pass if there is only a single shared index? An industrious and eager person might suggest raising a warning if a significant (10%, 50% or 90 %?) number of indices are discarded. Then again, not sure how relevant that is.

These kinds of comments would be fantastic in-line comments as part of a review! Would also make it easier to each. Therefore I will open them individually.

**Review by tjlane — COMMENTED** [review]
Flagging @cvass's three questions inline.

File: meteor/scale.py [hunk 1]
```diff
@@ -19,11 +19,16 @@
...
 
 DIMENSION_OF_MILLER_INDEX: int = 3
+_MAX_SCALE_FACTOR: float = 1e300
+_NON_FINITE_RESIDUAL_PENALTY: float = 1e30
 
 
 class ParameterLengthMismatchError(ValueError): ...
 
 
+class ScalingError(RuntimeError): ...
```
Comment (tjlane) [inline meteor/scale.py:30]: You define a ScalingError, but it only triggers if one of the two input datasets has a negative or nan mean. Doesn't this technically qualify as an Error for unclean data -> negative or nan structure factors detected?

File: meteor/scale.py [hunk 2]
```diff
@@ -165,50 +189,62 @@ def scale_maps(
...
-    sqrt_inverse_variance = 1.0 / np.sqrt(ref_variance + to_scale_variance)
+
+    valid = np.isfinite(ref_amps) & np.isfinite(to_amps)
+    if ref_sigmas is not None:
+        valid &= np.isfinite(ref_sigmas) & (ref_sigmas > 0.0)
+    if to_sigmas is not None:
+        valid &= np.isfinite(to_sigmas) & (to_sigmas > 0.0)
+
+    n_valid = int(valid.sum())
+    if n_valid == 0:
```
Comment (tjlane) [inline meteor/scale.py:210]: Log a warning here when `n_valid / len(valid)` falls below 0.5

**Comment by tjlane** [comment]
@cvazz you asked

> Is now the expected behavior that previous errors will not occur if this check is passed?

The answer is yes. If you can use this branch to run your example from #162, and ensure it does, that would be fantastic.

File: meteor/scale.py (hunk 1 above)
Comment (tjlane) [reply meteor/scale.py:30]: Not necessarily. I believe that most of the time this error will get raised as a result of bad inputs, but the truth is that other things could go wrong. The checks do not query the input per se, but the result of a scaling run -- so I think leaving the name general like this is the right move.

**Comment by cvazz** [comment]
On my dataset which is rather special, I now get a new error. I am not sure how general this error is/will be and it might be prudent to move this discussion outside of this pull request. But since you asked me to test it on the dataset that produced the error in #162, here we go:


Background to what/why I did this: Commonly missing reflections are set to zero but this is almost certainly wrong. The best estimate of this value lies in the model. So I wrote a function replacing any nan or missing value with the corresponding values of the model.

```
#create maps
struc = gemmi.read_structure(pdb_file)
map_dark_comp = gemmi_structure_to_calculated_map(struc, high_resolution_limit=dmin-.1)
map_dark_comp = cut_resolution(map_dark_comp, high_resolution_limit=dmin)
ds = rs.read_mtz(mtz_file)
map_mtz = rsmap.Map(ds, **columns, cell=ds.cell, spacegroup=ds.spacegroup)
map_mtz = cut_resolution(map_mtz, high_resolution_limit=dmin)


# double scaling
map_mtz = scale_maps(reference_map=map_dark_comp, map_to_scale=map_mtz)
map_filled = fill_na_with_model(map_to_fill=map_mtz, map_reference=map_dark_comp)
map_filled = scale_maps(reference_map=map_dark_comp, map_to_scale=map_filled)
```

Here it fails on the second scale maps call. It has nothing to do with either the specific values applied (i tried adding a global offset to all) or cutting out extra reflections (ensuring map_to_scale has no other reflections than those present in reference_map). I always get this error:
(Remaining code to recreate the example can be found below the error message print out)
```
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
Cell In[79], [line 31](vscode-notebook-cell:?execution_count=79&line=31)
     27 
     28 # double scaling
     29 map_mtz = scale_maps(reference_map=map_dark_comp, map_to_scale=map_mtz)
     30 map_filled = fill_na_with_model(map_to_fill=map_mtz, map_reference=map_dark_comp)
---> [31](vscode-notebook-cell:?execution_count=79&line=31) map_filled = scale_maps(reference_map=map_dark_comp, map_to_scale=map_filled)

File ~/anaconda3/envs/xtr-estimator/lib/python3.12/site-packages/meteor/scale.py:216, in scale_maps(reference_map, map_to_scale, scale_mode, weight_using_uncertainties, least_squares_loss)
    211     raise RuntimeError(msg)
    213 initial_scaling_parameters: ScaleParameters = (initial_c,) + (0.0,) * (
    214     scale_mode.number_of_parameters - 1
    215 )
--> [216](https://file+.vscode-resource.vscode-cdn.net/Users/sbielfel/Nextcloud2/time_resolved/occupancy-estimation/paperplots/~/anaconda3/envs/xtr-estimator/lib/python3.12/site-packages/meteor/scale.py:216) optimization_result = opt.least_squares(
    217     compute_residuals,
    218     initial_scaling_parameters,
    219     loss=least_squares_loss,
    220 )
    221 optimized_parameters: ScaleParameters = optimization_result.x
    223 optimized_scale_factors = compute_scale_factors(
    224     miller_indices=unmodified_map_to_scale.index,
    225     scale_parameters=optimized_parameters,
    226     scale_mode=scale_mode,
    227 )

File ~/anaconda3/envs/xtr-estimator/lib/python3.12/site-packages/scipy/_lib/_util.py:717, in _workers_wrapper.<locals>.inner(*args, **kwds)
    715 with MapWrapper(_workers) as mf:
    716     kwargs['workers'] = mf
--> [717](https://file+.vscode-resource.vscode-cdn.net/Users/sbielfel/Nextcloud2/time_resolved/occupancy-estimation/paperplots/~/anaconda3/envs/xtr-estimator/lib/python3.12/site-packages/scipy/_lib/_util.py:717)     return func(*args, **kwargs)

File ~/anaconda3/envs/xtr-estimator/lib/python3.12/site-packages/scipy/optimize/_lsq/least_squares.py:926, in least_squares(fun, x0, jac, bounds, method, ftol, xtol, gtol, x_scale, loss, f_scale, diff_step, tr_solver, tr_options, jac_sparsity, max_nfev, verbose, args, kwargs, callback, workers)
    922 def _dummy_hess(x, *args):
    923     # we don't care about Hessian evaluations
    924     return x
--> [926](https://file+.vscode-resource.vscode-cdn.net/Users/sbielfel/Nextcloud2/time_resolved/occupancy-estimation/paperplots/~/anaconda3/envs/xtr-estimator/lib/python3.12/site-packages/scipy/optimize/_lsq/least_squares.py:926) vector_fun = VectorFunction(
    927     fun_wrapped,
    928     x0,
    929     jac_wrapped,
    930     _dummy_hess,
    931     finite_diff_rel_step=diff_step,
    932     finite_diff_jac_sparsity=jac_sparsity,
    933     finite_diff_bounds=bounds,
    934     workers=workers
    935 )
    936 ###########################################################################
    938 f0 = vector_fun.fun(x0)

File ~/anaconda3/envs/xtr-estimator/lib/python3.12/site-packages/scipy/optimize/_differentiable_functions.py:614, in VectorFunction.__init__(self, fun, x0, jac, hess, finite_diff_rel_step, finite_diff_jac_sparsity, finite_diff_bounds, sparse_jacobian, workers)
    608     raise ValueError("Whenever the Jacobian is estimated via "
    609                      "finite-differences, we require the Hessian to "
    610                      "be estimated using one of the quasi-Newton "
    611                      "strategies.")
    613 self.fun_wrapped = _VectorFunWrapper(fun)
--> [614](https://file+.vscode-resource.vscode-cdn.net/Users/sbielfel/Nextcloud2/time_resolved/occupancy-estimation/paperplots/~/anaconda3/envs/xtr-estimator/lib/python3.12/site-packages/scipy/optimize/_differentiable_functions.py:614) self._update_fun()
    616 self.v = np.zeros_like(self.f)
    617 self.m = self.v.size

File ~/anaconda3/envs/xtr-estimator/lib/python3.12/site-packages/scipy/optimize/_differentiable_functions.py:707, in VectorFunction._update_fun(self)
    705 def _update_fun(self):
    706     if not self.f_updated:
--> [707](https://file+.vscode-resource.vscode-cdn.net/Users/sbielfel/Nextcloud2/time_resolved/occupancy-estimation/paperplots/~/anaconda3/envs/xtr-estimator/lib/python3.12/site-packages/scipy/optimize/_differentiable_functions.py:707)         self.f = self.fun_wrapped(xp_copy(self.x))
    708         self._nfev += 1
    709         self.f_updated = True

File ~/anaconda3/envs/xtr-estimator/lib/python3.12/site-packages/scipy/optimize/_differentiable_functions.py:424, in _VectorFunWrapper.__call__(self, x)
    422 def __call__(self, x):
    423     self.nfev += 1
--> [424](https://file+.vscode-resource.vscode-cdn.net/Users/sbielfel/Nextcloud2/time_resolved/occupancy-estimation/paperplots/~/anaconda3/envs/xtr-estimator/lib/python3.12/site-packages/scipy/optimize/_differentiable_functions.py:424)     return np.atleast_1d(self.fun(x))

File ~/anaconda3/envs/xtr-estimator/lib/python3.12/site-packages/scipy/optimize/_lsq/least_squares.py:263, in _WrapArgsKwargs.__call__(self, x)
    262 def __call__(self, x):
--> [263](https://file+.vscode-resource.vscode-cdn.net/Users/sbielfel/Nextcloud2/time_resolved/occupancy-estimation/paperplots/~/anaconda3/envs/xtr-estimator/lib/python3.12/site-packages/scipy/optimize/_lsq/least_squares.py:263)     return self.f(x, *self.args, **self.kwargs)

File ~/anaconda3/envs/xtr-estimator/lib/python3.12/site-packages/meteor/scale.py:194, in scale_maps.<locals>.compute_residuals(scale_parameters)
    190     msg += "Recommend: check the input data for severe outliers."
    191     raise RuntimeError(msg)
    193 difference_after_scaling = (
--> [194](https://file+.vscode-resource.vscode-cdn.net/Users/sbielfel/Nextcloud2/time_resolved/occupancy-estimation/paperplots/~/anaconda3/envs/xtr-estimator/lib/python3.12/site-packages/meteor/scale.py:194)     scale_factors * map_to_scale.amplitudes - reference_map.amplitudes
    195 )
    196 residuals = np.array(sqrt_inverse_variance * difference_after_scaling, dtype=np.float64)
    198 # filter NaNs in input -- are simply missing values

File ~/anaconda3/envs/xtr-estimator/lib/python3.12/site-packages/pandas/core/generic.py:2193, in NDFrame.__array_ufunc__(self, ufunc, method, *inputs, **kwargs)
   2189     @final
   2190     def __array_ufunc__(
   2191         self, ufunc: np.ufunc, method: str, *inputs: Any, **kwargs: Any
   2192     ):
-> [2193](https://file+.vscode-resource.vscode-cdn.net/Users/sbielfel/Nextcloud2/time_resolved/occupancy-estimation/paperplots/~/anaconda3/envs/xtr-estimator/lib/python3.12/site-packages/pandas/core/generic.py:2193)         return arraylike.array_ufunc(self, ufunc, method, *inputs, **kwargs)

File ~/anaconda3/envs/xtr-estimator/lib/python3.12/site-packages/pandas/core/arraylike.py:276, in array_ufunc(self, ufunc, method, *inputs, **kwargs)
    273 kwargs = _standardize_out_kwarg(**kwargs)
    275 # for binary ops, use our custom dunder methods
--> [276](https://file+.vscode-resource.vscode-cdn.net/Users/sbielfel/Nextcloud2/time_resolved/occupancy-estimation/paperplots/~/anaconda3/envs/xtr-estimator/lib/python3.12/site-packages/pandas/core/arraylike.py:276) result = maybe_dispatch_ufunc_to_dunder_op(self, ufunc, method, *inputs, **kwargs)
    277 if result is not NotImplemented:
    278     return result

File pandas/_libs/ops_dispatch.pyx:113, in pandas._libs.ops_dispatch.maybe_dispatch_ufunc_to_dunder_op()
--> [113](https://file+.vscode-resource.vscode-cdn.net/Users/sbielfel/Nextcloud2/time_resolved/occupancy-estimation/paperplots/pandas/_libs/ops_dispatch.pyx:113) 'Could not get source, probably due dynamically evaluated source code.'

File ~/anaconda3/envs/xtr-estimator/lib/python3.12/site-packages/pandas/core/ops/common.py:76, in _unpack_zerodim_and_defer.<locals>.new_method(self, other)
     72             return NotImplemented
     74 other = item_from_zerodim(other)
---> [76](https://file+.vscode-resource.vscode-cdn.net/Users/sbielfel/Nextcloud2/time_resolved/occupancy-estimation/paperplots/~/anaconda3/envs/xtr-estimator/lib/python3.12/site-packages/pandas/core/ops/common.py:76) return method(self, other)

File ~/anaconda3/envs/xtr-estimator/lib/python3.12/site-packages/pandas/core/arraylike.py:206, in OpsMixin.__rmul__(self, other)
    204 @unpack_zerodim_and_defer("__rmul__")
    205 def __rmul__(self, other):
--> [206](https://file+.vscode-resource.vscode-cdn.net/Users/sbielfel/Nextcloud2/time_resolved/occupancy-estimation/paperplots/~/anaconda3/envs/xtr-estimator/lib/python3.12/site-packages/pandas/core/arraylike.py:206)     return self._arith_method(other, roperator.rmul)

File ~/anaconda3/envs/xtr-estimator/lib/python3.12/site-packages/pandas/core/series.py:6154, in Series._arith_method(self, other, op)
   6152 def _arith_method(self, other, op):
   6153     self, other = self._align_for_op(other)
-> [6154](https://file+.vscode-resource.vscode-cdn.net/Users/sbielfel/Nextcloud2/time_resolved/occupancy-estimation/paperplots/~/anaconda3/envs/xtr-estimator/lib/python3.12/site-packages/pandas/core/series.py:6154)     return base.IndexOpsMixin._arith_method(self, other, op)

File ~/anaconda3/envs/xtr-estimator/lib/python3.12/site-packages/pandas/core/base.py:1391, in IndexOpsMixin._arith_method(self, other, op)
   1388     rvalues = np.arange(rvalues.start, rvalues.stop, rvalues.step)
   1390 with np.errstate(all="ignore"):
-> [1391](https://file+.vscode-resource.vscode-cdn.net/Users/sbielfel/Nextcloud2/time_resolved/occupancy-estimation/paperplots/~/anaconda3/envs/xtr-estimator/lib/python3.12/site-packages/pandas/core/base.py:1391)     result = ops.arithmetic_op(lvalues, rvalues, op)
   1393 return self._construct_result(result, name=res_name)

File ~/anaconda3/envs/xtr-estimator/lib/python3.12/site-packages/pandas/core/ops/array_ops.py:273, in arithmetic_op(left, right, op)
    260 # NB: We assume that extract_array and ensure_wrapped_if_datetimelike
    261 #  have already been called on `left` and `right`,
    262 #  and `maybe_prepare_scalar_for_op` has already been called on `right`
    263 # We need to special-case datetime64/timedelta64 dtypes (e.g. because numpy
    264 # casts integer dtypes to timedelta64 when operating with timedelta64 - GH#22390)
    266 if (
    267     should_extension_dispatch(left, right)
    268     or isinstance(right, (Timedelta, BaseOffset, Timestamp))
   (...)    271     # Timedelta/Timestamp and other custom scalars are included in the check
    272     # because numexpr will fail on it, see GH#31457
--> [273](https://file+.vscode-resource.vscode-cdn.net/Users/sbielfel/Nextcloud2/time_resolved/occupancy-estimation/paperplots/~/anaconda3/envs/xtr-estimator/lib/python3.12/site-packages/pandas/core/ops/array_ops.py:273)     res_values = op(left, right)
    274 else:
    275     # TODO we should handle EAs consistently and move this check before the if/else
    276     # (https://github.com/pandas-dev/pandas/issues/41165)
    277     # error: Argument 2 to "_bool_arith_check" has incompatible type
    278     # "Union[ExtensionArray, ndarray[Any, Any]]"; expected "ndarray[Any, Any]"
    279     _bool_arith_check(op, left, right)  # type: ignore[arg-type]

File ~/anaconda3/envs/xtr-estimator/lib/python3.12/site-packages/pandas/core/roperator.py:19, in rmul(left, right)
     18 def rmul(left, right):
---> [19](https://file+.vscode-resource.vscode-cdn.net/Users/sbielfel/Nextcloud2/time_resolved/occupancy-estimation/paperplots/~/anaconda3/envs/xtr-estimator/lib/python3.12/site-packages/pandas/core/roperator.py:19)     return right * left

File ~/anaconda3/envs/xtr-estimator/lib/python3.12/site-packages/reciprocalspaceship/dtypes/internals.py:469, in BaseMaskedArray.__array_ufunc__(self, ufunc, method, *inputs, **kwargs)
    466         return NotImplemented
    468 # for binary ops, use our custom dunder methods
--> [469](https://file+.vscode-resource.vscode-cdn.net/Users/sbielfel/Nextcloud2/time_resolved/occupancy-estimation/paperplots/~/anaconda3/envs/xtr-estimator/lib/python3.12/site-packages/reciprocalspaceship/dtypes/internals.py:469) result = maybe_dispatch_ufunc_to_dunder_op(
    470     self, ufunc, method, *inputs, **kwargs
    471 )
    472 if result is not NotImplemented:
    473     return result

File pandas/_libs/ops_dispatch.pyx:113, in pandas._libs.ops_dispatch.maybe_dispatch_ufunc_to_dunder_op()
--> [113](https://file+.vscode-resource.vscode-cdn.net/Users/sbielfel/Nextcloud2/time_resolved/occupancy-estimation/paperplots/pandas/_libs/ops_dispatch.pyx:113) 'Could not get source, probably due dynamically evaluated source code.'

File ~/anaconda3/envs/xtr-estimator/lib/python3.12/site-packages/pandas/core/ops/common.py:76, in _unpack_zerodim_and_defer.<locals>.new_method(self, other)
     72             return NotImplemented
     74 other = item_from_zerodim(other)
---> [76](https://file+.vscode-resource.vscode-cdn.net/Users/sbielfel/Nextcloud2/time_resolved/occupancy-estimation/paperplots/~/anaconda3/envs/xtr-estimator/lib/python3.12/site-packages/pandas/core/ops/common.py:76) return method(self, other)

File ~/anaconda3/envs/xtr-estimator/lib/python3.12/site-packages/pandas/core/arraylike.py:206, in OpsMixin.__rmul__(self, other)
    204 @unpack_zerodim_and_defer("__rmul__")
    205 def __rmul__(self, other):
--> [206](https://file+.vscode-resource.vscode-cdn.net/Users/sbielfel/Nextcloud2/time_resolved/occupancy-estimation/paperplots/~/anaconda3/envs/xtr-estimator/lib/python3.12/site-packages/pandas/core/arraylike.py:206)     return self._arith_method(other, roperator.rmul)

File ~/anaconda3/envs/xtr-estimator/lib/python3.12/site-packages/reciprocalspaceship/dtypes/internals.py:1278, in NumericArray._arith_method(self, other, op)
   1276     raise NotImplementedError("can only perform ops with 1-d structures")
   1277 if len(self) != len(other):
-> [1278](https://file+.vscode-resource.vscode-cdn.net/Users/sbielfel/Nextcloud2/time_resolved/occupancy-estimation/paperplots/~/anaconda3/envs/xtr-estimator/lib/python3.12/site-packages/reciprocalspaceship/dtypes/internals.py:1278)     raise ValueError("Lengths must match")
   1279 if not (is_float_dtype(other) or is_integer_dtype(other)):
   1280     raise TypeError("can only perform ops with numeric values")

ValueError: Lengths must match
```

```
from meteor import rsmap
import pandas as pd
def fill_na_with_model(
    map_to_fill: rsmap.Map,
    map_reference: rsmap.Map,
    amplitude_threshold: float = 1.0,
    high_resolution_limit: float = None,
) -> rsmap.Map:
    """Fill NaN values and missing reflections in `map_to_fill` using `map_reference`.

    Handles each of the three crystallographic columns individually
    (amplitudes F, uncertainties SIGF, phases PHI). Column names may differ
    between the two maps; the reference is translated into the target's
    column names before filling.

    If the reference map has no SIGF column, SIGF is estimated as 10% of F.
    Reference rows with F < `amplitude_threshold` are dropped before filling.
    """
    # ---------- 1. Resolve column names on both sides ----------
    f_col_tgt = map_to_fill.amplitude_column_name
    phi_col_tgt = map_to_fill.phase_column_name
    sigf_col_tgt = map_to_fill.uncertainties_column_name

    f_col_ref = map_reference.amplitude_column_name
    phi_col_ref = map_reference.phase_column_name

    # ---------- 2. Build a translated reference with target's column names ----------
    ref_translated = pd.DataFrame(index=map_reference.index)
    ref_translated[f_col_tgt] = map_reference[f_col_ref].values 
    ref_translated[phi_col_tgt] = map_reference[phi_col_ref].values

    if map_reference.has_uncertainties:
        sigf_col_ref = map_reference.uncertainties_column_name
        ref_translated[sigf_col_tgt] = map_reference[sigf_col_ref].values
    else:
        ref_translated[sigf_col_tgt] = 0.60 * map_reference[f_col_ref].values

    ref_translated = ref_translated[list(map_to_fill.columns)]

    # ---------- 2b. Drop low-amplitude reference rows ----------
    n_before = len(ref_translated)
    ref_translated = ref_translated[ref_translated[f_col_tgt] >= amplitude_threshold]
    ref_translated = ref_translated.drop(index=(0, 0, 0), errors="ignore")
    if high_resolution_limit is not None:
        ref_translated = ref_translated[map_reference.compute_dHKL() >= high_resolution_limit]  # type: ignore
    n_after = len(ref_translated)

    n_dropped = n_before - n_after
    if n_dropped:
        logger.warning(
            f"Dropped {n_dropped} reference reflections with "
            f"{f_col_tgt} < {amplitude_threshold}.",
            stacklevel=2,
        )

    # ---------- 3. Index bookkeeping ----------
    complete_rows = (
        map_to_fill.amplitudes.notna()
        & map_to_fill.phases.notna()
        & map_to_fill.uncertainties.notna()
    )
    fill_idx = set(map_to_fill.index[complete_rows])
    ref_idx = set(ref_translated.index)

    missing_in_fill = list(ref_idx - fill_idx)  # rows to add
    missing_in_ref = fill_idx - ref_idx  # rows we can't fill

    if missing_in_ref:
        logger.warning(
            f"Found {len(missing_in_ref)} indices in `map_to_fill` that are "
            f"missing from `map_reference` (or filtered out by threshold). "
            f"These cannot be filled.",
            stacklevel=2,
        )

    filled_map = map_to_fill.copy()
    plt.plot(1 / map_to_fill.compute_dHKL(), map_to_fill.amplitudes, ".")
    # ---------- 4. Add rows entirely missing from the target ----------
    if missing_in_fill:
        filled_map = pd.concat([filled_map, ref_translated.loc[missing_in_fill]])
        # Re-sort so the MultiIndex is lexsorted again — silences the
        # PerformanceWarning from .update() below and speeds up later .loc calls.
        filled_map = filled_map.sort_index()
        logger.info(f"Replaced {len(ref_translated.loc[missing_in_fill])} NaN values.")

    # ---------- 5. Fill NaN cells in pre-existing rows ----------
    filled_map.update(ref_translated, overwrite=False)
    filled_map.cell = (
        map_to_fill.cell
    )  # ensure cell info is preserved for later .loc calls
    filled_map.spacegroup = map_to_fill.spacegroup
    new_bits = filled_map.loc[missing_in_fill]
    # plt.yscale("log")
    # plt.xscale("log")
    # plt.ylabel("Amplitude")
    # plt.xlabel("1/distance (1/Å)")
    # plt.plot(1 / new_bits.compute_dHKL(), new_bits.amplitudes, ".")
    # plt.show()

    return filled_map

import gemmi
from meteor.sfcalc import gemmi_structure_to_calculated_map
from meteor.scale import scale_maps

from meteor.utils import cut_resolution
import reciprocalspaceship as rs
datadir = "/Users/sbielfel/Nextcloud2/time_resolved/data/o1_structures/"

config = apply_config_PL_3ps()
pdb_file=config["input_files"]["pdb_dark"]
if False:
    mtz_file = config["input_files"]["map_triggered"]
    columns = dict(amplitude_column="F", uncertainty_column="SIGF", phase_column="PHIF-model")
else:
    mtz_file = config["input_files"]["map_dark"]
    columns = dict(amplitude_column="F-obs-filtered", uncertainty_column="SIGF-obs-filtered", phase_column="PHIF-model")
dmin = 2.8

#create maps
struc = gemmi.read_structure(pdb_file)
map_dark_comp = gemmi_structure_to_calculated_map(struc, high_resolution_limit=dmin-.1)
map_dark_comp = cut_resolution(map_dark_comp, high_resolution_limit=dmin)
ds = rs.read_mtz(mtz_file)
map_mtz = rsmap.Map(ds, **columns, cell=ds.cell, spacegroup=ds.spacegroup)
map_mtz = cut_resolution(map_mtz, high_resolution_limit=dmin)

**Comment by tjlane** [comment]
@cvazz thanks for the detailed repro — I was able to reproduce the failure, and I think there are two separate things going on. tl;dr the PR does fix the failure mode from #162, but you're hitting a *different* bug with the same surface symptom.

## Heads up first

Your traceback is from `meteor/scale.py:194` doing `scale_factors * map_to_scale.amplitudes - reference_map.amplitudes` — that code path only exists in the **released** version, not on this PR branch. Your `xtr-estimator` env is still on PyPI meteor. Could you `pip install git+https://github.com/rs-station/meteor.git@tjlane/scale-error` and rerun? I expect the #162 traceback specifically will be gone.

## What I reproduced

Minimal repro that triggers the same `ValueError: Lengths must match` symptom on **both** the released code and this PR branch:

```python
import pandas as pd
from meteor.rsmap import Map
from meteor.scale import scale_maps

def make_map(hkls, amps):
    df = pd.DataFrame({'F': amps, 'PHI': [0.0]*len(amps), 'SIGF': [1.0]*len(amps)})
    df.index = pd.MultiIndex.from_tuples(hkls, names=['H','K','L'])
    return Map(df, cell=(10.,10.,10.,90.,90.,90.), spacegroup=1)

ref = make_map([(1,0,0),(0,1,0),(0,0,1),(1,1,0),(1,0,1)], [10.,8.,6.,4.,2.])
to  = make_map([(1,0,0),(1,0,0),(0,1,0),(0,0,1),(1,1,0),(1,0,1)], [5.,5.,4.,3.,2.,1.])
#                ^^^^^^^ duplicate HKL — mimics what fill_na_with_model produces
scale_maps(reference_map=ref, map_to_scale=to)
# -> ValueError: operands could not be broadcast together with shapes (5,) (6,)
```

## Root cause

Your `fill_na_with_model` does

```python
missing_in_fill = ref_idx - fill_idx          # ref minus *complete* rows of fill
filled_map = pd.concat([filled_map, ref_translated.loc[missing_in_fill]]).sort_index()
```

where `fill_idx = set(map_to_fill.index[complete_rows])`. So any HKL that *already exists* in `map_to_fill` but has a NaN in F/PHI/SIGF, **and** also exists in the reference, gets re-added by the `pd.concat` rather than filled in place. The returned `Map` has duplicate `(H,K,L)` rows. `scale_maps` then calls `filter_common_indices`, which does `df.loc[common_indices]` — that expands duplicates on one side but not the other (`index.intersection(...)` is unique), and the two sides go into the residual computation with mismatched lengths. Same surface symptom as #162, different upstream cause.

## Suggestions

Two independent fixes, both worth doing:

1. **In meteor** (this PR or a follow-up): detect non-unique input indices and raise `ScalingError` up front with an actionable message, instead of letting it surface as a broadcasting error.
2. **In your `fill_na_with_model`**: change `fill_idx = set(map_to_fill.index[complete_rows])` to `fill_idx = set(map_to_fill.index)` so that partially-NaN rows are filled in place via the `.update(...)` call further down, instead of being re-added by `concat`. Alternatively `filled_map = filled_map[~filled_map.index.duplicated(keep='first')]` after the concat.

I'm happy to add the early uniqueness check + a regression test to this branch if @tjlane agrees — it would convert your opaque error into a clear `ScalingError("input maps must have unique HKL indices")`, and is independent of the #162 fix.

Could you also run a quick sanity check after switching to the PR branch — drop the duplicates from `map_filled` (`map_filled = map_filled[~map_filled.index.duplicated()]`) before the second `scale_maps` call, and confirm it succeeds end-to-end? That isolates whether your dataset has any *other* edge case the PR doesn't cover.

**Comment by cvazz** [comment]
Oh man, thanks for asking claude haha. It works now!

I had seen the duplicates issue, I had written `map_filled.drop_duplicates()` and I ran the test in both versions, the main and the one of your branch. Well, maybe I forgot to write `inplace=True`.

In any case, i guess it is finally time to approve :)

File: meteor/scale.py (hunk 2 above)
Comment (cvazz) [reply meteor/scale.py:210]: Thanks for incorporating

**Review by cvazz — APPROVED** [review]
